url47af6b1535_rolandotr_db_comparison_tgz-pJ8-methodology_HistoryJ8/src/utils/Optimization.java

```diff
@@ -0,0 +1,332 @@
+					State s = states[i].clone();
+					s.moveToBitsExchangedChild();
+					newStates.add(s);
+				}*/
+				if (root.hasSizeOfMessageChildren()){
+					State s = states[i].clone();
+					s.moveToSizeOfMessagesChild();
+					newStates.add(s);
+				}
+				if (root.hasCryptoCallsChildren()){
+					State s = states[i].clone();
+					s.moveToCryptoCallsChild();
+					newStates.add(s);
+				}
+				if (root.hasMemoryChildren()){
+					State s = states[i].clone();
+					s.moveToMemoryChild();
+					newStates.add(s);
+				}
+				if (root.hasFinalSlowPhaseChildren()){
+					State s = states[i].clone();
+					s.moveToFinalSlowPhaseChild();
+					newStates.add(s);
+				}
+			}
+		}
+		State[] result = new State[newStates.size()];
+		newStates.toArray(result);
+		search(root, protocols, attributes, result);
+	}
+	
+	private static void search(Node root, DBProtocol[] protocols, Attribute[] attributes){
+		List<DBProtocol> tmp;
+		Hashtable<String, String> names;
+		BestProtocol best = new BestProtocol();
+		best.setNumberOfRounds(root.getTotalBitsExchangedDuringFastPhase()/2);
+		if (!best.dominate(root, attributes)){
+			//entonces nadie podra dominar a este hombre!
+			tmp = new LinkedList<>();
+		}
+		else tmp = getNonDpominated(root, protocols, attributes);
+		names = getProtocolNames(tmp);
+		//if (tmp.size() > 0){
+		if (names.size() == 1){
+			System.out.println("One solution found, which is");
+			History.printInfo(root, attributes);
+			for (DBProtocol p : tmp) {
+				//History.printInfo(p, attributes);
+				System.out.print(p.getIdentifier()+", ");
+			}
+			return;
+		}
+		State clone = root.getState().clone();
+		if (root.hasMafiaChildren()){
+			root.getState().moveToMafiaChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		if (root.hasDistanceChildren()){
+			root.getState().moveToDistanceChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		if (root.hasTerroristChildren()){
+			root.getState().moveToTerroristChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		/*if (root.hasBitsExchangedChildren()){
+			root.getState().moveToBitsExchangedChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}*/
+		if (root.hasSizeOfMessageChildren()){
+			root.getState().moveToSizeOfMessagesChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		if (root.hasCryptoCallsChildren()){
+			root.getState().moveToCryptoCallsChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		if (root.hasMemoryChildren()){
+			root.getState().moveToMemoryChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+		if (root.hasFinalSlowPhaseChildren()){
+			root.getState().moveToFinalSlowPhaseChild();
+			search(root, protocols, attributes);
+			root.setState(clone);
+		}
+	}
+
+	private static void searchByIteration(Node root, DBProtocol[] protocols, 
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

The cyclomatic complexity of this method is 16. By comparison, 99% of the methods in the CodeGuru reference dataset have a lower cyclomatic complexity. This indicates the method has a high number of decisions and it can make the logic difficult to understand and test. We recommend that you simplify this method or break it into multiple methods.